feat(parquet): add variant encoder/decoder

sfc-gh-mbojanczyk opened the pull request:
### Rationale for this change
This adds a basic Variant encoder/decoder to start the process of supporting the new [Variant encoding spec](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md) in the Apache Go Parquet library. Variants are useful for efficiently storing and accessing data, especially in things like Iceberg tables.

### What changes are included in this PR?
This adds logic to encode and decode Variants, but does not yet plumb that logic through to either Arrow or Parquet. The PR's getting beefy as is, and this seems to be a good standalone unit to get feedback on.

Still to implement are the handling of decimal primitives.

For ease of implementation, the Metadata keys are only stored in unsorted order. This makes the creation of an encoded Variant simpler as one can serialize data as its being added. For sorted Metadata keys to work, you'd need to buffer data and only create objects at the very end so that the appropriate width of indicies can be chosen.

### Are these changes tested?
There are unit tests throughout to test that marshaling produces the expected binary output as per the spec, and to ensure that unmarshaling can spit out the expected values. There are many levels of unit tests, from testing individual marshaling bits to testing the marshaling and unmarshaling of entire Variants.

### Are there any user-facing changes?
With this PR, no. This is simply a library to create Variants, but does not plumb the output into Parquet or Arrow.

zeroshade commented on parquet/variants/util.go:
```diff
@@ -0,0 +1,154 @@
+
+// Reads a little-endian encoded uint (betwen 1 and 8 bytes wide) from a raw buffer at a specified
+// offset and returns its value. If any part of the read would be out of bounds, this returns an error.
+func readUint(raw []byte, offset, size int) (uint64, error) {
+	if size < 1 || size > 8 {
+		return 0, fmt.Errorf("invalid size, must be in range [1,8]: %d", size)
+	}
+	if maxPos := offset + size; maxPos > len(raw) {
+		return 0, fmt.Errorf("out of bounds: trying to access position %d, max position is %d", maxPos, len(raw))
+	}
+	var ret uint64
+	for i := range size {
+		ret |= uint64(raw[i+offset]) << (8 * i)
+	}
+	return ret, nil
+}
```
`binary.LittlEndian.Uint64(raw[offset:size])` ?

zeroshade replied:
Then you could probably just do something like:

```go
switch size {
case 1:
    var val uint8
    _, err := binary.Decode(raw[offset:size], binary.LittleEndian, &val)
    return val, err
case 2:
    var val uint16
    _, err := binary.Decode(raw[offset:size], binary.LittleEndian, &val)
    return val, err
case 4:
    var val uint32
    _, err := binary.Decode(raw[offset:size], binary.LittleEndian, &val)
    return val, err
case 8:
    var val uint64
    _, err := binary.Decode(raw[offset:size], binary.LittleEndian, &val)
    return val, err
}
```

right?

sfc-gh-mbojanczyk replied:
The subtlety here: you need to be able to decode things of widths 1-8 _inclusive_. For example, the [field offset size](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md#value-header-for-object-basic_type2) of an object can be 1, 2, 3, or 4 bytes. `binary.Decode` can handle the 1, 2, and 4 case, but cannot handle the 3-byte width case (see [this example](https://go.dev/play/p/5QKZ1ove5Ab) in the Go playground).

For the odd-width cases I suppose we could pad the buffer with additional zeros, but that's getting gnarly just to make `binary.Decode` work IMO

sfc-gh-mbojanczyk replied:
Ah, I see where your confusion is coming from (the spec doesn't make it easy to find all these in one place).

There are _so many places_ where widths are set, `is_large` is solely one place. In this case `is_large` indicates whether `num_elements` in the [Object Data](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md#value-data-for-object-basic_type2) is 1 or 4 bytes. I'm talking about `field_id` and `field_offset` sizes. From the header documentation:

> `field_offset_size_minus_one` and `field_id_size_minus_one` are 2-bit values that represent the number of bytes used to encode the field offsets and field ids.

In this case, `field_id` and `field_offset` can be 1-4 bytes.

Perhaps a better example here is in the [header](https://github.com/apache/parquet-format/blob/master/VariantEncoding.md#value-header-for-short-string-basic_type1) for `Short String`:

> When `basic_type` is 1, value_header is a 6-bit `short_string_header` ... The `short_string_header` value is the length of the string.

So a short string's length can be represented by 1-6 bytes inclusive.

sfc-gh-mbojanczyk replied:
(Sorry, misread the Short String header, it's six bits. Removed that example)

zeroshade commented on parquet/variants/util.go:
```diff
@@ -0,0 +1,154 @@
+// Checks that a given range is in the provided raw buffer.
+func checkBounds(raw []byte, low, high int) error {
+	maxPos := len(raw)
+	if low >= maxPos {
```
we should be also validating that both values are >= 0 and that high >= low

sfc-gh-mbojanczyk replied:
Good call. Done, plus I realized that I forgot tests for this function, so I added that as well.

zeroshade commented on parquet/variants/util.go:
```diff
@@ -0,0 +1,154 @@
+
+// Encodes a number of a specified width in little-endian format and writes to a writer.
+func encodeNumber(val int64, size int, w io.Writer) {
+	buf := make([]byte, size)
+	for i := range size {
+		buf[i] = byte(val)
+		val >>= 8
+	}
+	w.Write(buf)
```
why not use the `encoding/binary` functions to do this instead? such as `binary.LittleEndian.Put*` or `binary.LittleEndian.Append*` etc.

sfc-gh-mbojanczyk replied:
Similar to the comment in `readUint()`, the `binary` package only provides Put/Append functionality for `uint{16,32,64}`, but we need it for the whole range from 1-8 bytes.

zeroshade replied:
you can use `binary.Encode(buf, binary.LittleEndian, val)` which will handle the whole range for you

sfc-gh-mbojanczyk replied:
Similar vein to `Decode()`- this will work for the powers-of-two widths, but if we've got a width that's not (ie. 3) we'd encode with additional padding since we've got to encode to the next power of two width.

This isn't the end of the world (encoding over the minimal necessary width _is_ within spec), so I have fewer objections here. I'd argue with `Decode()` rolling our own is more of a necessity due to not being in control of what comes in (eg. the Java library _can_ encode 3-byte wide numbers, so it's gotta be handled). This keeps the encode and decode logic fairly similar. Also, FWIW, I feel like the intention of the spec authors is to minimize the number of bytes used to encode things, see the existence of `Short String` even though a primitive `String` type exists.

zeroshade replied:
The problem with rolling our own here is that we're going to need to manage the big/little endian logic ourselves then so that this runs properly on big-endian systems.

zeroshade commented on parquet/variants/util.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// Returns the basic type the passed in value should be encoded as, or undefined if it cannot be handled.
+func kindFromValue(val any) BasicType {
```
we should cache the conversions somewhere if possible, reflection is slow so let's avoid duplicating calls to it if we don't need to

sfc-gh-mbojanczyk replied:
Hmm, this is an interesting point, I need to chew on this a bit.

My gut instinct is that this isn't terribly cacheable, the only place where we _could_ potentially cache is for elements in a slice/array (assuming we don't have nested slices) or for elements in a map (ditto). NB, this also only gets hit in the `builder.Marshal()` codepath- one can also totally build things with the provided builders (`Primitive(), Object() and Array()`), and avoid reflection entirely.

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	switch val := v.(type) {
+	case bool:
+		return marshalBoolean(val, w), nil
+	case int:
+		return marshalInt(int64(val), w), nil
+	case int8:
+		return marshalInt(int64(val), w), nil
+	case int16:
+		return marshalInt(int64(val), w), nil
+	case int32:
+		return marshalInt(int64(val), w), nil
```
why marshal  everything as `int64`? Why not marshal them as their actual type via generics?

zeroshade replied:
do you mind if I push some changes to the branch with a possible solution?

sfc-gh-mbojanczyk replied:
Please do! I'm not territorial about this :)

zeroshade replied:
I didn't delete any of your code, but rather commented it out where I replaced it with some generic stuff.

My personal opinion is that we should encode based on the type they give us rather than trying to downcast for them. If they want it to be smaller and use fewer bytes, they can downcast it themselves :smile:

Anyway, let me know what you think of the changes I made

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+		return marshalFloat(val, w), nil
+	case float64:
+		return marshalDouble(val, w), nil
+	case string:
+		if allOpts&MarshalAsUUID != 0 {
+			return marshalUUID([]byte(val), w), nil
+		}
+		return marshalString(val, w), nil
```
can we accept `uuid.UUID` instead?

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	switch val := v.(type) {
+	case bool:
+		return marshalBoolean(val, w), nil
+	case int:
```
what about uints?

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	case primitiveNull:
+		dest.Set(reflect.Zero(dest.Type()))
+	case primitiveTrue, primitiveFalse:
+		if kind != reflect.Bool && kind != reflect.Interface {
```
why `kind != reflect.Interface`?

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(iv))
```
I don't understand why you're checking and using `reflect.Interface` here. Can you explain? For reference, `reflect.Interface` just means the type is *an* interface, it doesn't mean it is the empty interface (i.e. `interface{}` or `any`)

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+				return fmt.Errorf("int value of %d will overflow dest", iv)
+			}
+			switch kind {
+			case reflect.Int:
```
what about the uint types?

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveString:
+		str, err := unmarshalString(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.String || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(str))
+		} else if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 {
+			dest.Set(reflect.ValueOf([]byte(str)))
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveBinary:
+		bytes, err := unmarshalBinary(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(bytes))
+		} else if kind == reflect.String {
+			dest.Set(reflect.ValueOf(string(bytes)))
+		} else {
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
```
this could be simplified with a generic most likely

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+			return fmt.Errorf("cannot decode Variant value of %s into dest %s", prim, kind)
+		}
+	case primitiveUUID:
+		bytes, err := unmarshalUUID(raw, offset)
+		if err != nil {
+			return err
+		}
+		if kind == reflect.Slice && dest.Type().Elem().Kind() == reflect.Uint8 || kind == reflect.Interface {
+			dest.Set(reflect.ValueOf(bytes))
+		} else if kind == reflect.String {
+			dest.Set(reflect.ValueOf(string(bytes)))
+		} else {
+			return fmt.Errorf("cannot decode Variant UUID into dest %s", kind)
+		}
```
what if they used uuid.UUID? or `[16]byte`?

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	return nil
+}
+
+func marshalNull(w io.Writer) int {
+	hdr, _ := primitiveHeader(primitiveNull)
+	w.Write([]byte{hdr})
+	return 1
+}
```
the return for all of these should probably be (int, error) and just do `return w.Write(...)`

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	hdr, _ := primitiveHeader(primitiveUUID)
+	w.Write([]byte{hdr})
+
+	// A UUID is 16 bytes. Either pad or truncate to this length.
+	if len(uuid) > 16 {
+		uuid = uuid[:16]
+	} else if pad := 16 - len(uuid); pad > 0 {
+		uuid = append(uuid, make([]byte, pad)...)
+	}
```
a uuid must be exactly 16 bytes, we should error if it is not exactly 16 bytes

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	return 17
+}
+
+func unmarshalUUID(raw []byte, offset int) ([]byte, error) {
+	if err := checkBounds(raw, offset, offset+17); err != nil {
+		return nil, err
+	}
+	return raw[offset+1 : offset+17], nil
+}
```
use `[16]byte` or `uuid.UUID` please

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
```
use `unsafe.String` to ensure we don't perform an unnecessary copy here.

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+		if endIdx > maxPos {
+			return "", fmt.Errorf("end index is out of bounds: trying to access position %d, max position is %d", endIdx, maxPos)
+		}
+		return string(raw[offset+1 : endIdx]), nil
```
use `unsafe.String` to avoid an unnecessary copy

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,678 @@
+	return 9
+}
+
+// func decodeTimestamp(raw []byte, offset int) (int64, error) {
+// 	ts, err := readUint(raw, offset+1, 8)
+// 	if err != nil {
+// 		return -1, err
+// 	}
+// 	return int64(ts), nil
+// }
```
remove commented out code?

zeroshade reviewed (commented):
Made an initial pass on this and left a bunch of comments. I'll try to take another look later on

sfc-gh-mbojanczyk commented:
> Made an initial pass on this and left a bunch of comments. I'll try to take another look later on

Thanks so much! I'll get to addressing the comments here shortly (was out for a few days)

sfc-gh-mbojanczyk reviewed (commented):
Mostly addressed your comments here, added some context for the others. The biggest diff IMO is using `uuid.UUID` now, which definitely simplifies some stuff around there.

I've also moved this review out of draft. Figure it's in pretty decent reviewable shape as-is.

Thanks for the first pass- looking forward to polishing this up!

zeroshade commented on parquet/variants/primitive.go:
```diff
@@ -0,0 +1,674 @@
+	case int32:
+		return marshalInt(int64(val), w), nil
+	case int64:
+		if allOpts&MarshalAsTime != 0 {
```
instead of needing this option, why not just use `arrow.Time32` or `arrow.Time64`? if they want a Time they should pass that.

sfc-gh-mbojanczyk commented:
> @sfc-gh-mbojanczyk Hey, sorry for the long delay here. I've gone through multiple iterations and discussions with people, plus pulling inspiration from the C++, Spark, and parquet-java implementations to figure out a good design. Please let me know what you think of the updated version! Thanks!

No worries- I got caught in a whirlwind over here myself and was about to dust this off too :) Lemme take a peek here after the long weekend.

lidavidm reviewed (approved):
I like this, this is very clean and straightforward

zeroshade commented on parquet/variant/variant.go:
```diff
@@ -0,0 +1,725 @@
+)
+
+func basicTypeFromHeader(hdr byte) BasicType {
+	return BasicType(hdr & basicTypeMask)
```
turns out there's no need for a `debug.Assert` here, because we're doing `& basicTypeMask` we're guaranteed the result is going to be in the correct range for the `BasicType` enum: `[0,3]`

lidavidm commented on parquet/variant/variant_test.go:
Thanks. I wonder if fuzzing the variants specifically would be a more manageable case to start with.

zeroshade commented on parquet/variant/variant.go:
```diff
@@ -0,0 +1,722 @@
+type PrimitiveType int
+
+const (
+	PrimitiveInvalid            PrimitiveType = iota - 1 // Unknown
```
These constants have specific values defined by the spec, so the zero value needs to be Null as the constant for "Null" type is 0. I don't really have wiggle room to change that. Though, technically if all variant building is done through the builder, we could probably get away without exporting this enum and these constants.

zeroshade commented on parquet/variant/builder.go:
```diff
@@ -0,0 +1,852 @@
+func (b *Builder) Build() (Value, error) {
+	nkeys := len(b.dictKeys)
+	totalDictSize := 0
+	for _, k := range b.dictKeys {
```
Mostly just out of convenience and saving the couple bytes in the size of the builder itself (which is negligible honestly). There's no reason I can't keep a running tally in AddKey, I'll make that change.

zeroshade commented on parquet/variant/basic_type_string.go:
sure, i'll update the go-generate lines and have it rename the files.

zeroshade commented on parquet/variant/doc.go:
```diff
@@ -0,0 +1,142 @@
+// For direct construction of complex nested structures, you can use the low-level
+// methods:
+//
+//	var b variant.Builder
```
>  Things also get tricky with sorting of metadata keys and building (as you can't know the index of the dictionary key until build time, which then changes an Object's index, which can then change the size of offsets, etc...).

Well that's an entirely different ball of worms. Even if you use the nested object approach, things are still tricky if you want to sort the metadata keys as you end up having to patch, modify, or entirely rewrite any already written objects when you change the field index by sorting the metadata keys. (See the [rust impl](https://github.com/apache/arrow-rs/pull/7452/files#diff-83c4572d83951bc1234fd12d3cc6537b1e025188f06c39b378b0f7b9c32aa4f3)). So using the nested approach with multiple buffers doesn't help us make sorting metadata keys more efficient.

> I don't think it can be done

yea, that's why the API looks the way it does right now, so that we can build everything using a single buffer instead of needing multiple buffers and copying a bunch of stuff around. 

I've managed to sketch something out that enables using parent/child builders to manage the nested building while still requiring that the object be built in order. Kind of a compromise between the approaches, I'll try to clean it up a bit and see if it's worth pushing it as an option.

> Once again, up to you. At least with child builders the parent builder can keep track of all nested things, and either error out if its children haven't called Finish(), or the parent can call each child's Finish()

Yea, there's definitely benefits to having the parent builders to keep track of the nested children in terms of usability. Just trying to figure out if the code complexity is worth it.

zeroshade replied:
Okay, what do you think of something like this:

```go
	var b variant.Builder

	obj, err := b.NewObject()
        // handle err

	obj.AppendInt("id", 1)

	observe, err := obj.AppendObject("observation")
        // handle err

	observe.AppendString("location", "In the Volcano")
	observe.AppendString("time", "12:34:56")

	value, err := observe.AppendObject("value")
	// handle err

	value.AppendInt("humidity", 456)
	value.AppendInt("temperature", 123)
	value.Finish()
	observe.Finish()

	species, err := obj.AppendObject("species")
	// handle err
	species.AppendString("name", "lava monster")
	species.AppendInt("population", 6789)
	species.Finish()

	obj.Finish()
        v, err := b.Build()
	// handle err
	out, err := json.Marshal(v)	
	/* result is this `{
		"id": 1,
		"observation": {
			"location": "In the Volcano",
			"time": "12:34:56",
			"value": {
				"humidity": 456,
				"temperature": 123
			}
		},
		"species": {
			"name": "lava monster",
			"population": 6789
		}
	}` */
```

sfc-gh-mbojanczyk replied:
I like it. I assume you'll do some handling if the parent has `Finish()` called without the children's `Finish()` (either panic, or just call their `Finish()` methods, right?

lidavidm replied:
(Also, presumably we can add the higher level builders separately/on top of this API? Does it have to be built into the builder for this first pass?)

zeroshade replied:
> I like it. I assume you'll do some handling if the parent has Finish() called without the children's Finish() (either panic, or just call their Finish() methods, right?

Yea, `Finish()` errors if the children haven't had `Finish()` called first.

> You can also still just directly append a map or slice instead of keeping track manually - would that solve your use case? (Also, presumably we can add the higher level builders separately/on top of this API? Does it have to be built into the builder for this first pass?)

I think @lidavidm has a good point here. The "compromise" example I put together is just built on top of the existing Builder API I've put together by creating the "higher-level" API on top of the "lower-level" one that I've already built. Any objections to pushing through what we've already got here and then following up with the higher-level builder that handles the offsets and fields for you as a separate PR afterwards @sfc-gh-mbojanczyk?

zeroshade replied:
> At the end of the day, this is your codebase and I'm just tossing in my 2c.

Yup, but I still value the feedback. Especially as I am often *way* too deep in the weeds and will miss things that can affect usability by consumers. So the feedback is important.

>  `Append()` does the overwhelming majority of the heavy lifting for users and is probably going to be the most used entry point, so this PR is fully featured enough (ie. it's a good hole for folk to drop into).

Cool. I'll push this as is for now and then we can look at creating the higher-level "nicer" API for simplifying object/array construction later on. Another reason is that I want to get the basic variant support out and my "compromise" version has a lot of duplicated code in there that I wasn't able to get to my liking yet.

sfc-gh-mbojanczyk commented:
Can't actually toss in my approval since I'm the original author, but this PR looks good to me.
